drift/core/beamtransfer.py

```diff
@@ -831,190 +831,264 @@ def _generate_svdfiles(self, regen=False, skip_svd_inv=False):
+                    nmodes, self.telescope.num_pol_sky, self.telescope.lmax + 1
+                )
 
-                    if not skip_svd_inv:
+                if not skip_svd_inv:
+                    # Find the pseudo-inverse of the beam matrix and save to disk.
+                    # First try la.pinv, which uses a least-squares solver.
+                    try:
                         ibeam = la.pinv(beam)
-                        # Find the pseudo-inverse of the beam matrix and save to disk.
-                        dset_ibsvd[fi, :, :, :nmodes] = ibeam.reshape(
-                            self.telescope.num_pol_sky, self.telescope.lmax + 1, nmodes
+                    except la.LinAlgError as e:
+                        # If la.pinv fails, try la.pinv2, which is SVD-based and
+                        # more likely to succeed. If successful, add file attribute
+                        # indicating pinv2 was used for this frequency.
+                        print(
+                            "***pinv failure: m = %d, fi = %d. Trying pinv2..."
+                            % (mi, fi)
                         )
+                        try:
+                            ibeam = la.pinv2(beam)
+                            if "inv_bsvd_from_pinv2" not in fs.attrs.keys():
+                                fs.attrs["inv_bsvd_from_pinv2"] = [fi]
+                            else:
+                                bad_freqs = fs.attrs["inv_bsvd_from_pinv2"]
+                                fs.attrs["inv_bsvd_from_pinv2"] = bad_freqs.append(fi)
+                        except:
+                            # If pinv2 fails, close SVD and beam files gracefully,
+                            # then print error message
+                            fs.close()
+                            fm.close()
+                            raise Exception("pinv2 failure: m = %d, fi = %d" % (mi, fi))
+
+                    dset_ibsvd[fi, :, :, :nmodes] = ibeam.reshape(
+                        self.telescope.num_pol_sky, self.telescope.lmax + 1, nmodes
+                    )
 
-                    # Save out the singular values for each block
-                    dset_sig[fi, :nmodes] = sig
-
-            # Write a few useful attributes.
-            fs.attrs["baselines"] = self.telescope.baselines
-            fs.attrs["m"] = mi
-            fs.attrs["frequencies"] = self.telescope.frequencies
+                # Save out the singular values for each block
+                dset_sig[fi, :nmodes] = sig
 
-            fs.close()
-            fm.close()
+        # Write a few useful attributes.
+        fs.attrs["baselines"] = self.telescope.baselines
+        fs.attrs["m"] = mi
+        fs.attrs["frequencies"] = self.telescope.frequencies
 
-        # If we're part of an MPI run, synchronise here.
-        mpiutil.barrier()
+        # We completed all the computations successfully, so set file's
+        # complete attribute accordingly.
+        fs.attrs["complete"] = True
```

Comment:
I'm not sure about the pattern of a `complete` attribute. Seems a little hackish. You could try the `lock_file` manager from https://github.com/radiocosmology/caput/blob/master/caput/misc.py#L134

At least that way you'd guarantee that if a file is present with the right way, then it's complete.

Reply:
I'm not sure if `lock_file` will work here: it appears to only work with the file if the MPI rank is 0, but we are assigning different m's to different ranks, so we actually want each rank to be able to open and write to its own file.

(If I'm reading things correctly, the docstring for `lock_file` is wrong: it states that the rank==0 restriction will only be enforced if `comm is not None`, but the routine still checks `mpiutil.rank0` if `comm is None`.)

Reply:
Yeah good point. I agree that it won't work as it currently stands. We could fix `lock_file` to match the documented behaviour and use it, but that's a little annoying to need to do that. I leave the decision up to you.

Reply:
Okay, I submitted a separate caput PR to fix that, and implemented `lock_file` in my new commits below